File: backend/app/models/product.py
"""
SQLAlchemy model for products.
Aligned with 1_database_schema.sql.
"""

import uuid
from datetime import datetime, timezone
from decimal import Decimal

from sqlalchemy import (
    Boolean,
    DateTime,
    ForeignKey,
    Integer,
    JSON,
    Numeric,
    String,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class Product(Base):
    __tablename__ = "products"
    __table_args__ = (
        UniqueConstraint("workspace_id", "sku", name="uq_products_ws_sku"),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    workspace_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("workspaces.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    category_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        # FK to product_categories deferred
        nullable=True,
    )
    type: Mapped[str] = mapped_column(
        String(50), default="physical", nullable=False
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    sku: Mapped[str | None] = mapped_column(String(100), nullable=True)
    unit_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        # FK to units_of_measure deferred
        nullable=True,
    )
    base_price: Mapped[Decimal] = mapped_column(
        Numeric(10, 2), nullable=False
    )
    cost_price: Mapped[Decimal] = mapped_column(
        Numeric(10, 2), default=Decimal("0.00"), nullable=False
    )
    tax_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        # FK to taxes deferred
        nullable=True,
    )
    min_stock_alert: Mapped[int] = mapped_column(Integer, default=5)
    dynamic_attributes: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    is_deleted: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )

File: backend/app/Models/Product.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Concerns\HasUuids;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

/**
 * @property string $id
 * @property string $workspace_id
 * @property string|null $category_id
 * @property string $type          physical|service|digital|subscription
 * @property string $name
 * @property string|null $sku
 * @property string|null $unit_id
 * @property float $base_price
 * @property float $cost_price
 * @property string|null $tax_id
 * @property int|null $min_stock_alert
 * @property array|null $dynamic_attributes
 * @property bool $is_deleted
 */
class Product extends Model
{
    use HasUuids;

    protected $table = 'products';
    protected $keyType = 'string';
    public $incrementing = false;

    protected $fillable = [
        'workspace_id',
        'category_id',
        'type',
        'name',
        'sku',
        'unit_id',
        'base_price',
        'cost_price',
        'tax_id',
        'min_stock_alert',
        'dynamic_attributes',
        'is_deleted',
    ];

    protected function casts(): array
    {
        return [
            'base_price' => 'decimal:2',
            'cost_price' => 'decimal:2',
            'min_stock_alert' => 'integer',
            'dynamic_attributes' => 'array',
            'is_deleted' => 'boolean',
        ];
    }

    // ── Relationships ──────────────────────────────────────────

    public function category(): BelongsTo
    {
        return $this->belongsTo(ProductCategory::class, 'category_id');
    }

    // ── Scopes ─────────────────────────────────────────────────

    public function scopeActive($query)
    {
        return $query->where('is_deleted', false);
    }
}
